strapi/strapi
# Fix Document Service Typing Issues

### What does it do?

- Fix document service return types
- Add various type documentation

### Why is it needed?

- Currently, the document service is not dynamically reusing the methods’ parameters in their return types, resulting in wrong types.

### How to test it?

- Go to a typescript application with generated types
- Use the document service to make several queries (making use of properties like `fields` or `populate`)
- Example
```typescript
  // restaurant has draft & publish enabled, the publish method is available
  await strapi.documents('api::restaurant.restaurant').publish({
    documentId: 'iuhzf-hbzefyg-567vzef',
    locale: 'fr',
    filters: { name: { $startsWith: 'Restaurant' } },
  });

  const entries = await strapi
    .documents('api::restaurant.restaurant')
    .findMany({ fields: 'name,shortDescription', populate: 'categories' });

  for (const restaurant of entries) {
    // Only name, shortdescription and categories are select in the query. They should be the only ones available in the types
   // (+ documentId and id that are always there)
    const { name, shortDescription, categories } = restaurant;

    console.log(
      name,
      shortDescription,
      categories.map((cat) => cat.name)
    );
  }
```
- Make sure the results make sense and that the types are correctly inferred.

### Related issue(s)/PR(s)

Partially fix #20735
